doc/developer/design/20230209_invalid_accumulations.md

```diff
@@ -0,0 +1,119 @@
+
+At a high level, our solution approach is to add a post-processing operator immediately after each reduction operator that evaluates if an error occurred during reduction and then produces an error in the appropriate error stream if so. This design avoids changing reduction operators to potentially create error streams, thus reducing the complexity of the change. Two important aspects need to be tackled for this solution approach to work: (1) We need to find a method to concretely implement the post-processing operator; (2) We need to devise a strategy to externalize the information that an error occurred in the output of the reduction operator.
+
+Regarding point (1), the post-processing operator could be implemented as a special-purpose operator in rendering or as a map-project chain with the relevant error processing code encoded as instances of `MirScalarExpr`. The former option maintains all the implementation complexity within rendering. An argument can be made against this choice, however. Positioning the post-processing operator after a reduction implies that the operator will take the arrangement at the reduction's output, process its rows, and output a collection. Thus, arrangement reuse in downstream operators, e.g., joins, is hindered. This lack of arrangement reuse is a serious concern due to both the overhead in re-arranging the data downstream as well as the increase in memory footprint for a query. By contrast, the latter option embraces error handling from invalid accumulations in reductions as a semantic, query-level construct, but it may suffer from lower expression evaluation performance than rendering-level alternatives. Importantly, since MIR transformations eventually reason about error-freeness of expressions, it is necessary to introduce the map-project chain at a higher level of abstraction. We argue, in particular, for introducing this enriched reduce translation at the level of HIR-to-MIR lowering. At this point in the query optimization pipeline, all aggregate functions have already been appropriately detected and collected into `Reduce` nodes. Thus, we do not need to reason about query nesting or how to rewrite multiple interrelated aggregates with different placements in a query (e.g., consider different aggregates put in a `SELECT` and a `HAVING` clause).
```

Comment:
Regarding the first question: The idea is to change the semantics of a `Reduce` output in the case of invalid accumulations in a way that has to be uniformly implemented across all operators in rendering. So that way, we do not need to know about the `ReductionType` at a higher-level of abstraction. Each rendered operator detects invalid accumulations in its specific way, but needs to report the errors in a uniform way.

Regarding the second question: With the `map-project` approach, we give the optimizer a chance to hoist these `map`s and `projection`s, so that arrangements can be exposed out of reductions. If we introduce a post-processing operator after we apply transforms, then the operator will not be moved and we are stuck with a collection.

Regarding the third point: It is a very good observation, thanks! I added some more discussion in a new Risk Management section in the document.